src/cargo/ops/cargo_output_depinfo.rs

```diff
@@ -0,0 +1,148 @@
+        None => path,
+        Some(base) => match path.strip_prefix(base) {
+            Ok(relpath) => relpath,
+            _ => path,
+        }
+    };
+    relpath.to_str().ok_or(internal("path not utf-8")).map(ToOwned::to_owned)
+}
+
+fn read_dep_file<P: AsRef<Path>>(path: P) -> CargoResult<DepFile> {
+    let mut file = File::open(&path).map_err(|_|
+        human("error opening ".to_string() + path.as_ref().to_str().unwrap_or("(bad unicode")))?;
+    let mut contents = String::new();
+    let _ = file.read_to_string(&mut contents)?;
+    let mut spl = contents.split('\0');
+    let dir = spl.next().ok_or(internal("dependency file empty"))?;
+    let dep_txt = spl.next().ok_or(internal("dependency file missing null byte"))?;
+    let mut result = Vec::new();
+    for line in dep_txt.lines() {
+        let mut line_spl = line.split(": ");
+        if let Some(target) = line_spl.next() {
+            if let Some(deps) = line_spl.next() {
+                let deps = deps.split_whitespace().map(ToOwned::to_owned).collect();
+                result.push(DepLine {
+                    target: target.to_string(),
+                    deps: deps,
+                });
+            }
+        }
+    }
+    Ok(DepFile {
+        dir: dir.to_string(),
+        deps: result,
+    })
+}
+
+fn add_deps(depfile: &DepFile, deps: &mut HashSet<PathBuf>) {
+    let dep_dir = PathBuf::from(&depfile.dir);
+    for depline in &depfile.deps {
+        for dep in &depline.deps {
+            deps.insert(dep_dir.join(dep));
+        }
+    }
+}
+
+// TODO: this should be based on CompileFilter
+fn match_target(target: &Target, example: Option<&str>, is_lib: bool)
+    -> Option<(&'static str, bool)>
+{
+    match *target.kind() {
+        TargetKind::Lib(..) => Some(("lib", is_lib)),
+        TargetKind::Bin => {
+            if example.is_none() && !is_lib {
+                Some(("bin", true))
+            } else {
+                None
+            }
+        }
+        TargetKind::Example => {
+            if let Some(e) = example {
+                if e == target.name() {
+                    return Some(("example", true));
+                }
+            }
+            None
+        }
+        _ => None
+    }
+}
+
+fn target_filename(context: &mut Context, unit: &Unit) -> CargoResult<PathBuf> {
+    let (dir, base) = context.link_stem(&unit).ok_or(internal("can't get link stem"))?;
+    if unit.target.is_lib() {
+        Ok(dir.join(["lib", &base, ".rlib"].concat()))
+    } else {
+        Ok(dir.join(base))
+    }
+}
+
+pub fn output_depinfo(path: &str, ws: &Workspace, config: &Config,
+                      build_config: BuildConfig) -> CargoResult<()> {
+    let (packages, resolve) = ops::resolve_ws(&ws)?;
+    let profiles = ws.profiles();
+    let profile = if build_config.release {
+        &profiles.release
+    } else {
+        &profiles.dev
+    };
+    let mut context = Context::new(&ws, &resolve, &packages, &config, build_config, &profiles)?;
```

Comment:
Perhaps this could be moved into `cargo_rustc` to avoid creating multiple contexts? It's not expensive or anything but it tends to be subtle enough to introduce lots of bugs, so deduplicating the logic would be best.